core/shared/src/main/scala/zio/ZRef.scala

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright 2017-2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio
+import java.util.concurrent.atomic.AtomicReference
+import scala.annotation.tailrec
+
+/**
+ * A mutable atomic reference for the `IO` monad. This is the `IO` equivalent of
+ * a volatile `var`, augmented with atomic operations, which make it useful as a
+ * reasonably efficient (if low-level) concurrency primitive.
+ *
+ * {{{
+ * for {
+ *   ref <- Ref.make(2)
+ *   v   <- ref.update[A0 <: A](_ + 3)
+ *   _   <- console.putStrLn("Value = " + v) // Value = 5
+ * } yield ()
+ * }}}
+ */
+abstract class ZRef[+E, -A, +B] extends Serializable { self =>
+  import ZRef._
+
+  /**
+   * Reads the value from the `ZRef`.
+   */
+  def get: IO[E, B]
+
+  /**
+   * Atomically modifies the `ZRef` with the specified function, which computes
+   * a return value for the modification. This is a more powerful version of
+   * `update`.
+   */
+  def modify[X]: ZRefModify[E, X, A]
```

Comment:
@regiskuckaertz could you explain why the intermediate classes are needed? From a cursory look, it seems like they wouldn't affect inference or quantification :-)

Reply:
That was my last attempt 😆 I tried:
```
def modify[A0 <: A, X](f: A => (X, A0))
def modify[A0 >: A, X](f: A0 => (X, A))
def modify[A0 >: A, A1 <: A, X](f: A0 => (X, A1))
```
but these don't work for the same reasons.

Reply:
Oh, the variance is the issue. And it works because the intermediates are invariant in the type parameters.

I think the previous encoding of `Managed` had that same issue and we had to introduce an abstract type member. So that's one direction you could look into; maybe that could help.

I might have time tonight/tomorrow to check out the branch and try some things if you don't solve it by then :-)

Reply:
I thought that would be something like this. I tried but failed to find the shape of that abstract type but I'd be super glad if you could help me on that line of reasoning 🙏